presto-main/src/main/java/com/facebook/presto/operator/aggregation/PrecisionRecallAggregation.java

```diff
@@ -0,0 +1,175 @@
+            state.setFalseWeights(new FixedNumericHistogram((int) (numBins), 0.0, 1.0));
+        }
+
+        if (pred < 0 || pred > 1 || weight < 0.0) {
+            state.setNull(true);
+            return;
+        }
+
+        if (outcome) {
+            state.getTrueWeights().add(pred, weight);
+        }
+        else {
+            state.getFalseWeights().add(pred, weight);
+        }
+    }
+
+    @InputFunction
+    public static void input(
+            @AggregationState PrecisionRecallState state,
+            @SqlType(StandardTypes.BIGINT) long numBins,
+            @SqlType(StandardTypes.BOOLEAN) boolean outcome,
+            @SqlType(StandardTypes.DOUBLE) double pred)
+    {
+        PrecisionRecallAggregation.input(state, numBins, outcome, pred, PrecisionRecallAggregation.DEFAULT_WEIGHT);
+    }
+
+    @CombineFunction
+    public static void combine(
+            @AggregationState PrecisionRecallState state,
+            @AggregationState PrecisionRecallState otherState)
+    {
+        state.setNull(state.getNull() || otherState.getNull());
+        if (state.getNull()) {
+            return;
+        }
+
+        if (state.getTrueWeights() == null && otherState.getTrueWeights() != null) {
+            state.setTrueWeights(otherState.getTrueWeights().copy());
+            state.setFalseWeights(otherState.getFalseWeights().copy());
+            return;
+        }
+        if (state.getTrueWeights() != null && otherState.getTrueWeights() != null) {
+            state.getTrueWeights().mergeWith(otherState.getTrueWeights());
+            state.getFalseWeights().mergeWith(otherState.getFalseWeights());
+        }
+    }
+
+    @OutputFunction("map(double,double)")
+    public static void output(@AggregationState PrecisionRecallState state, BlockBuilder out)
+    {
+        if (state.getNull()) {
+            out.appendNull();
+            return;
+        }
+
+        SortedMap<Double, Double> result;
+        if (state.getTrueWeights() == null) {
+            result = PrecisionRecallAggregation.getEmptyResultMap();
+        }
+        else {
+            result = PrecisionRecallAggregation.getResultMap(state);
+        }
+
+        BlockBuilder entryBuilder = out.beginBlockEntry();
+        Iterator<Map.Entry<Double, Double>> resultIt = result.entrySet().iterator();
+        while (resultIt.hasNext()) {
+            final Map.Entry<Double, Double> entry = resultIt.next();
+            final Double recall = entry.getKey();
+            final Double precision = entry.getValue();
+            DoubleType.DOUBLE.writeDouble(entryBuilder, recall);
+            DoubleType.DOUBLE.writeDouble(entryBuilder, precision);
+        }
+        out.closeEntry();
+    }
+
+    private PrecisionRecallAggregation() {}
+
+    private static Double safeDiv(Double nom, Double denom)
+    {
+        return denom == 0.0 ? 1.0 : nom / denom;
+    }
+
+    private static SortedMap<Double, Double> getResultMap(@AggregationState PrecisionRecallState state)
```

Comment:
So if we go with the iterator approach I mentioned above, we can change this method to not return a map and directly write out the curve, to avoid repackaging in a map which we'll then just iterate over to serialize.

Reply:
I actually like using your idea of an iterator here as well, just a composite (zip) iterator.